### ISSUE TITLE
Explore richer quick-pick support

### ISSUE THREAD

**Issue by chrmarti** [Microsoft/vscode#45589]

**Comment by chrmarti** [comment]
Collecting a few recordings of current workarounds in [OneNote](https://microsoft.sharepoint.com/teams/DD_OTP/_layouts/OneNote.aspx?id=%2Fteams%2FDD_OTP%2FDocuments%2FTicino%2FNotebooks%2FVSCode&wd=target%28Development.one%7C8557A876-0489-401B-A073-CECF1B859862%2FRicher%20Quick-Pick%20Support%7C0F3D5FFD-CC91-8348-88E4-0256BE8BDCE1%2F%29).

Some obvious issue:

- Select subscriptions in Azure Account extension:
  - Flickers because the Quick-Pick is reopened after each selection.
  - Lost scroll state because Quick-Pick is reopened.
  - No 'OK' button to confirm.
  - No 'Cancel' button to bail.
  - No native checkboxes.

- Create & deploy website in App Service extension:
  - Flicker between steps.
  - No 'OK' button to confirm.
  - No 'Cancel' button to bail.
  - No 'Back' button to reconsider previous step.
  - Lost state when closing QuickPick inadvertently or for research.
  - Message boxes appear bottom right of the window, other steps are top center.

**Comment by chrmarti** [comment]
Moqups:
- [multi-select-moqup.pdf](https://github.com/Microsoft/vscode/files/1807514/multi-select-moqup.pdf)
- [create-website-moqup.pdf](https://github.com/Microsoft/vscode/files/1807516/create-website-moqup.pdf)

**Comment by eamodio** [comment]
Any chance https://github.com/Microsoft/vscode/issues/45869 can be included in the investigation/exploration?

**Comment by chrmarti** [comment]
@eamodio We'll keep that in mind, but it won't make the high priority list at this time.

Everyone: Here are some variations on the single item picker in the multi step flow:
- The first one comes with just 'Back' and 'New' buttons at the bottom, omitting 'Next' and 'Cancel' which can be reached by selecting an item with a left mouse click or the Enter key and clicking outside the picker or the Escape key.
- The second one improves on the actions as list items by separating them as their own 'group'.
- The third and fourth show the actions as action bar items at the top in different positions.

I also include the previous variants at the end for reference.

[single-item-picker-in-multi-step-flow.pdf](https://github.com/Microsoft/vscode/files/1824139/single-item-picker-in-multi-step-flow.pdf)

**Comment by eamodio** [comment]
@chrmarti Thanks.

My 2c &mdash; I much prefer the 2nd option (grouped actions) as it is easier to navigate with the keyboard and feels more extensible. At the same time, the 3rd option is also interesting, especially if it can be mixed with the 2nd option (if desired).

It seems like option 2 adds support to the quick pick for grouping, while option 3 adds support for a *navigation* group. Both of which I can see being useful (and also very similar in to menu commands).

**Comment by EricJizbaMSFT** [comment]
A few things:
1. I love groupings for quick picks. That being said, I think we also need to consider inputBoxes and groupings wouldn't really help in that situation. For example, the first thing I do after selecting "New Resource Group..." is enter the new name in an input box. I think having a "back" button like in the 1st or 3rd options would make the most sense for both input boxes and quick picks (although I don't think the "new" button is necessary for input boxes)
1. Not a fan of "New" as the bottom right button in option 1. I'm just too used to that location being the "confirm" button and so I feel like people would click it all the time as if it were the "next" button without reading the name

**Comment by chrmarti** [comment]
Good points. It has been pointed out that one issue with 1) is that the buttons change vertical position when the list is filtered and when moving between steps. That makes 3) and maybe 4) the preferred options for now. I will do some more mockups.

**Comment by chrmarti** [comment]
Going with the toolbar to the right of the filter field for now. Here are two more sets of mockups: Both show the 3 input types (single-select, multi-select and single-line text) after each other as they might appear in a multi-step input sequence. The second one adds a title area that holds the title the number of steps and the toolbar.
- [multi-step-with-toolbar.pdf](https://github.com/Microsoft/vscode/files/1830151/multi-step-with-toolbar.pdf)
- [multi-step-with-title.pdf](https://github.com/Microsoft/vscode/files/1830153/multi-step-with-title.pdf)

**Comment by chrmarti** [comment]
I haven't spent much time thinking about the API yet. You would probably pass the actions with each step, with the back and forward actions being given. The number of steps could be updated along the way or omitted.

**Comment by HoboKristian** [comment]
Would it also be possible to make it more decorable? Something like adding a small color-indicator next to items.

**Comment by chrmarti** [comment]
@HoboKristian That is very likely beyond the scope of this issue. Please check if any of the issues listed above in the description cover your case and create a new issue if not. Thanks.

**Comment by ArthurMa1978** [comment]
Would it support cross extension navigation? like extension A invoke a command of extension B, then the 'back' Button / Action can pull back to extension A.

**Comment by chrmarti** [comment]
Implemented the multi-select picker and did mockups for multi-step input for March. Closing as this concludes the exploration. Will track additional work in new issues.

**Comment by chrmarti** [comment]
@ArthurMa1978 Cross-extension navigation might be possible with the close cooperation of both extensions. But I don't have a draft of the API yet, so it is hard to tell for sure at the moment.

**Comment by ArthurMa1978** [comment]
@chrmarti got it, let's check once the API is ready
